File: flexdocs-go/internal/encryption/encryption.go
package encryption

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/sha512"
	"crypto/subtle"
	"encoding/hex"
	"fmt"
	"io"
	"os"
	"strings"

	"golang.org/x/crypto/pbkdf2"
)

const (
	ivLength    = 16
	keyLength   = 32
	iterations  = 100000
	saltDefault = "flexdocs-salt-v1"
)

var (
	derivedKey []byte
)

func init() {
	key := os.Getenv("ENCRYPTION_KEY")
	if key == "" {
		key = os.Getenv("NEXTAUTH_SECRET")
	}
	if key == "" {
		panic("ENCRYPTION_KEY or NEXTAUTH_SECRET must be set")
	}
	salt := os.Getenv("ENCRYPTION_SALT")
	if salt == "" {
		salt = saltDefault
	}
	derivedKey = pbkdf2.Key([]byte(key), []byte(salt), iterations, keyLength, sha512.New)
}

// Encrypt encrypts plaintext using AES-256-GCM. Format: iv:tag:ciphertext (hex)
func Encrypt(plaintext string) string {
	if plaintext == "" {
		return plaintext
	}

	block, err := aes.NewCipher(derivedKey)
	if err != nil {
		panic(err)
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		panic(err)
	}

	iv := make([]byte, gcm.NonceSize())
	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
		panic(err)
	}

	ciphertext := gcm.Seal(nil, iv, []byte(plaintext), nil)
	encrypted := ciphertext[:len(ciphertext)-gcm.Overhead()]
	tag := ciphertext[len(ciphertext)-gcm.Overhead():]

	return fmt.Sprintf("%s:%s:%s", hex.EncodeToString(iv), hex.EncodeToString(tag), hex.EncodeToString(encrypted))
}

// Decrypt decrypts ciphertext. Returns original string if not encrypted.
func Decrypt(ciphertext string) string {
	if ciphertext == "" {
		return ciphertext
	}

	parts := strings.SplitN(ciphertext, ":", 3)
	if len(parts) != 3 {
		return ciphertext // Not encrypted
	}

	iv, err := hex.DecodeString(parts[0])
	if err != nil {
		return ciphertext
	}
	tag, err := hex.DecodeString(parts[1])
	if err != nil {
		return ciphertext
	}
	encrypted, err := hex.DecodeString(parts[2])
	if err != nil {
		return ciphertext
	}

	block, err := aes.NewCipher(derivedKey)
	if err != nil {
		return ciphertext
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return ciphertext
	}

	// GCM expects ciphertext+tag combined
	combined := append(encrypted, tag...)
	plaintext, err := gcm.Open(nil, iv, combined, nil)
	if err != nil {
		return ciphertext // Decryption failed, return as-is
	}

	return string(plaintext)
}

// HashForSearch creates a deterministic HMAC hash for searching encrypted fields
func HashForSearch(plaintext string) string {
	h := hmac.New(sha256.New, derivedKey)
	h.Write([]byte(strings.ToLower(strings.TrimSpace(plaintext))))
	return hex.EncodeToString(h.Sum(nil))
}

// HashAPIKey creates a SHA-256 hash of an API key
func HashAPIKey(key string) string {
	h := sha256.Sum256([]byte(key))
	return hex.EncodeToString(h[:])
}

// VerifyAPIKey compares an API key against its hash (constant-time)
func VerifyAPIKey(key, hash string) bool {
	keyHash := HashAPIKey(key)
	return subtle.ConstantTimeCompare([]byte(keyHash), []byte(hash)) == 1
}

File: src/lib/encryption.ts
import crypto from 'crypto';

const ALGORITHM = 'aes-256-gcm';
const IV_LENGTH = 16;
const KEY_LENGTH = 32;
const ITERATIONS = 100000;

function getEncryptionKey(): Buffer {
  const secret = process.env.ENCRYPTION_KEY;
  if (!secret) {
    throw new Error('ENCRYPTION_KEY must be set — refusing to derive key from other secrets');
  }
  const salt = process.env.ENCRYPTION_SALT || 'flexdocs-salt-v1';
  return crypto.pbkdf2Sync(secret, salt, ITERATIONS, KEY_LENGTH, 'sha512');
}

export function encrypt(plaintext: string): string {
  if (!plaintext) return plaintext;
  const key = getEncryptionKey();
  const iv = crypto.randomBytes(IV_LENGTH);
  const cipher = crypto.createCipheriv(ALGORITHM, key, iv);

  let encrypted = cipher.update(plaintext, 'utf8', 'hex');
  encrypted += cipher.final('hex');

  const tag = cipher.getAuthTag();

  // Format: iv:tag:ciphertext (all hex)
  return `${iv.toString('hex')}:${tag.toString('hex')}:${encrypted}`;
}

export function decrypt(ciphertext: string): string {
  if (!ciphertext) return ciphertext;
  // Legacy plaintext or non-standard values are returned as-is for backward compat
  const parts = ciphertext.split(':');
  if (parts.length !== 3) return ciphertext;

  try {
    const key = getEncryptionKey();
    const iv = Buffer.from(parts[0], 'hex');
    const tag = Buffer.from(parts[1], 'hex');
    const encrypted = parts[2];

    const decipher = crypto.createDecipheriv(ALGORITHM, key, iv);
    decipher.setAuthTag(tag);

    let decrypted = decipher.update(encrypted, 'hex', 'utf8');
    decrypted += decipher.final('utf8');

    return decrypted;
  } catch {
    // Surface key-rotation or tampering issues instead of leaking ciphertext
    throw new Error('Failed to decrypt value: invalid ciphertext or encryption key mismatch');
  }
}

export function hashForSearch(plaintext: string): string {
  // Deterministic hash for searching encrypted fields
  const key = getEncryptionKey();
  return crypto.createHmac('sha256', key).update(plaintext.toLowerCase().trim()).digest('hex');
}
